### PR TITLE
1. Fetch UeContext based on SUPI

### PR DESCRIPTION
**Pull request by vthiruveedula**
2. Associate grpc channel to Ran Context during INIT itelf
3. Fetch Ran based on GnbIp and add RanUe to RanUeList

### CODE REVIEW

File: context/context.go [hunk 1]
```diff
@@ -265,7 +266,18 @@ func (context *AMFContext) AmfUeFindBySupi(supi string) (ue *AmfUe, ok bool) {
...
+	} else if context.EnableDbStore {
+		ue, ok = DbFetchUeBySupi(supi)
+		if ue != nil && ok {
+			logger.ContextLog.Infof("Ue with supi found in DB : ", supi)
+			context.UePool.Store(ue.Supi, ue)
+		} else {
+			logger.ContextLog.Infoln("Ue with Supi not found locally and in DB: ", supi)
+		}
+	} else {
+		logger.ContextLog.Infoln("Ue with Guti not found : ", supi)
```
Comment (badhrinathpa) [inline context/context.go:278]: Change statement to Ue with Supi not found.

File: context/amf_ue.go [hunk 2]
```diff
@@ -302,12 +302,21 @@ func (ue *AmfUe) UnmarshalJSON(data []byte) error {
...
-		ue.State[index].Set(fsm.StateType(states))
+		ue.State[index] = fsm.NewState(fsm.StateType(states))
 		if ue.RanUe[index] == nil {
 			ue.RanUe[index] = &RanUe{}
 		}
 		ue.RanUe[index].Log = logger.NgapLog.WithField(logger.FieldAmfUeNgapID, fmt.Sprintf("AMF_UE_NGAP_ID:%d", ue.RanUe[index].AmfUeNgapId))
+		ue.RanUe[index].RanUeNgapId = aux.RanUeNgapId
+		if ran != nil {
+			ran.RanUeList = append(ran.RanUeList, ue.RanUe[index])
+			ue.RanUe[index].Ran = ran
```
Comment (badhrinathpa) [inline context/amf_ue.go:318]: I think we update the ran after we fetch the UE/RANUE. so we need to see how that affects the ranUelist append part.

File: ngap/dispatcher.go [hunk 3]
```diff
@@ -23,34 +24,48 @@ import (
...
+		var ok bool
+		ran, ok = amfSelf.AmfRanFindByGnbId(sctplbMsg.GnbId)
+		if !ok {
+			logger.NgapLog.Infof("Create a new NG connection for: %s", sctplbMsg.GnbId)
+			ran = amfSelf.NewAmfRanId(sctplbMsg.GnbId)
+			ran.Amf2RanMsgChan = Amf2RanMsgChan
+			fmt.Println("DispatchLb, Create new Amf RAN ", sctplbMsg.GnbId)
+		}
+	} else if sctplbMsg.GnbIpAddr != "" {
+		fmt.Printf("GnbIpAddress received but no GnbId")
```
[the remark this answers is not in the document]
Comment (vthiruveedula) [reply ngap/dispatcher.go:43]: gnbip comes only for ngsetuprequest, subsequent messages goes with gnbid

File: context/amf_ue.go [hunk 4]
```diff
@@ -443,6 +459,8 @@ func (ue *AmfUe) init() {
 	ue.onGoing[models.AccessType__3_GPP_ACCESS] = new(OnGoing)
 	ue.onGoing[models.AccessType__3_GPP_ACCESS].Procedure = OnGoingProcedureNothing
 	ue.ReleaseCause = make(map[models.AccessType]*CauseAll)
+	ue.AmfInstanceName = os.Getenv("HOSTNAME")
```
Comment (badhrinathpa) [inline context/amf_ue.go:462]: do we need default values for these variables ?

File: nas/handler.go [hunk 5]
```diff
@@ -30,6 +33,26 @@ func HandleNAS(ue *context.RanUe, procedureCode int64, nasPdu []byte) {
 		ue.AmfUe = nas_security.FetchUeContextWithMobileIdentity(nasPdu)
 		if ue.AmfUe == nil {
 			ue.AmfUe = amfSelf.NewAmfUe("")
+		} else {
+			/* checking the guti-ue belongs to this amf instance */
+			id, _ := amfSelf.Drsm.FindOwnerInt32ID(int32(ue.AmfUe.Tmsi))
+			if id != nil && id.PodName != os.Getenv("HOSTNAME") {
+				rsp := &sdcoreAmfServer.AmfMessage{}
+				rsp.VerboseMsg = "Hello From AMF Pod !"
```
Comment (badhrinathpa) [inline nas/handler.go:41]: verbose msg could match the msg type

File: producer/1 [hunk 6]
```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 free5GC.org
+//
+// SPDX-License-Identifier: Apache-2.0
+//
```
Comment (badhrinathpa) [inline producer/1:4]: is this a test file ?

File: util/init_context.go [hunk 7]
```diff
@@ -18,6 +19,18 @@ import (
 	"github.com/omec-project/openapi/models"
 )
 
+func InitDrsm() (*drsm.Drsm, error) {
+	podname := os.Getenv("HOSTNAME")
+	podip := os.Getenv("POD_IP")
+	podId := drsm.PodId{PodName: podname, PodIp: podip}
+	opt := &drsm.Options{ResIdSize: 16, Mode: drsm.ResourceClient}
+	db := drsm.DbInfo{Url: "mongodb://mongodb-arbiter-headless",
```
Comment (badhrinathpa) [inline util/init_context.go:27]: can we use configured db url ?

File: ngap/handler.go [hunk 8]
```diff
@@ -1475,6 +1506,25 @@ func HandleInitialUEMessage(ran *context.AmfRan, message *ngapType.NGAPPDU) {
 				ranUe.Log.Warnf("Unknown UE [GUTI: %s]", guti)
 			} else {
 				ranUe.Log.Tracef("find AmfUe [GUTI: %s]", guti)
+				/* checking the guti-ue belongs to this amf instance */
+				id, _ := amfSelf.Drsm.FindOwnerInt32ID(int32(amfUe.Tmsi))
```
Comment (badhrinathpa) [inline ngap/handler.go:1510]: can convert this to a function to reuse. can do later.

File: context/db.go [hunk 9]
```diff
@@ -68,15 +69,26 @@ func SetupAmfCollection() {
...
+	logger.ContextLog.Infof("MondbName: %v, Url: %v", factory.AmfConfig.Configuration.AmfDBName, mongoDbUrl)
+
 	if Namespace != "" {
 		AmfUeDataColl = Namespace + "." + AmfUeDataColl
 	}
-
-	MongoDBLibrary.SetMongoDB(factory.AmfConfig.Configuration.AmfDBName, mongoDbUrl)
+	for {
+		MongoDBLibrary.SetMongoDB(factory.AmfConfig.Configuration.AmfDBName, mongoDbUrl)
+		if MongoDBLibrary.Client == nil {
```
Comment (badhrinathpa) [inline context/db.go:85]: we have a new function which returns the client to you I think. Its better to use that than directly access variable in library.

**Review by ajaythakurintel — APPROVED** [review]

**Review by thakurajayL — DISMISSED** [review]

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**